### ISSUE TITLE
clangd (vscode): Goto definition performs symlink resolution

### ISSUE THREAD

**Issue by decimad** [clangd/clangd#544]
Hi, at times I'm using symlinks within my source tree, if I need to make consistent changes to a module used in multiple projects.

The setup resemble something like this:
`
parent_folder
|- workspace_root
|  |- libs
|     |- module (symlink to module in parent folder)
|
module (destination)
`

Now when clangd is indexing the projects, everything seems to work fine. When I open the source files within the workspace, they have their "local paths" and clangd shows them parsed them nicely. However, if I use functions such as "go to definition", I end up in the resolved paths (parent_folder/module/..) instead of (parent_folder/workspace_root/libs/module/...), at which point clangd also will not work anymore (probably as this represenation of the path to the file is not part of the compilation database).

**Comment by HighCommander4** [comment]
Possibly related: #413, #197

**Comment by sam-mccall** [comment]
This is true, clangd treats symlinks as multiple names for the same file and generally asks the OS to canonicalize.

Without canonicalizing filenames, we don't know which of several symlinked paths to foo.h "go-to-definition" should jump to.

And there are setups where the symlink is certainly the incorrect path (e.g. synthesized by the build system) and the resolved name is the correct one.

So I'm sympathetic to the problem, but it's not clear how to address it in a way that's reasonable to maintain and doesn't break other equally-important use cases.

**Comment by ArcsinX** [comment]
It looks like `--path-mappings` option might help in your case.

```
--path-mappings=<parent_folder>\workspace_root\libs\module=<parent_folder>\module
```

**Comment by decimad** [comment]
Thank you for bringing this up, but I'm afraid the --path-mappings does not help in this regard, I also swapped the sides desperately :)
If that is of importance: The compilation database generated by cmake contains the uncanonicalized paths.

I will gladly settle with this issue, once I understand the other side of the story: I can't currently quite follow why the canonicalized path would be of importance, when the point of symlinks is that you want to transparently pretend something is somewhere where in fact it is not. If someone "inside" this transparency is canonicalizing around the symlink, then surely he is breaking the user's intent, no? (Sorry for my misunderstanding, I'm not so proficient with symlinks yet, as I'm an old windows user :D )

**Comment by ArcsinX** [comment]
> Thank you for bringing this up, but I'm afraid the --path-mappings does not help in this regard, I also swapped the sides desperately :)
  
I successfully tried the following (based on your description):
**Symlink**
```
D:\work\workspace_root\libs>mklink /J module D:\work\module
Junction created for module <<===>> D:\work\module
```

**D:\work\module\module.h**
```cpp
#pragma once
int module_init() { return 0; }
```
**D:\work\workspace_root\test.cpp**
```cpp
#include "module.h"
int test() {
  return module_init();
}
```

**compile_commands.json**
```json
[
{
  "directory": "D:/work/workspace_root/",
  "command": "g++.exe -I D:/work/workspace_root/libs/module/ -c D:/work/workspace_root/test.cpp",
  "file": "D:/work/workspace_root/test.cpp"
}
]
```

Extra `clangd` option:
`--path-mappings=D:\work\workspace_root\libs\module=D:\work\module` (in settings.json it should looks as `"--path-mappings=D:\\work\\workspace_root\\libs\\module=D:\\work\\module"`

Go-to-definition to `module_init()` in `test.cpp` jumps to `D:\work\workspace_root\libs\module\module.h` (as expected)

P.S.
`--path-mappings` does not work well on Windows because of case-sensitive strings compares.  
In this scenario drive letter should be *upper case* (but there are other scenarios when drive letter in `--path-mappings` should be lower case :( )

**Comment by decimad** [comment]
Right, I have up too early, I'm sorry. Once I crafted the source path correctly, the inludes could be resolved again, withouth hopping to the canonicalized patsh. However, now whenever I navigate to a declaration/definition within "test.cpp" itself, a second editor of "text.cpp" with the canonicalized paths pops up in vscode.

Also since recently, when the background indexer is indexing any file "some.cpp", it will also try to index "some.cpp," (note the appended comma), which fails with diagnostics.

**Comment by wwm0609** [comment]
If you use Clion instead of VsCode as the IDE, none of the above paranoid things would happen, cherrs!

However, Clion might have its own problems too, e.g not suitable for large projects as the UI shall freeze for a very long period frequently while you are coding

**Comment by lixin-wei** [comment]
`--path-mappings` is not an ideal way in my case.

In my project, the `module` path keeps being changed by my build tool. And I need to modify `--path-mappings` every time I rebuilt the project. Which is very frastrating.

I'm wondering whether there is any flag that can make clangd jump to uncanonicalized path directly.

**Comment by mckaydavis** [comment]
For those of you using VSCode with the `vscode-clangd` extension, I recommend the solution in https://github.com/clangd/vscode-clangd/pull/707

It adds a `clangd.followSymlinks` config option.  This fixes goto definition for project built in a Bazel sandbox that I was encountering this issue on. 

Thank you @cgrinker!

**Comment by JVApen** [comment]
Can this behavior be rechecked in a new version of VS Code? We use a sym link to all our generated code (such that different cmake presets use the same directory for the generated code, this is useful for index files). One of my colleagues pointed out that after opening a file via the sym link, it always refers to the actual path. (So switching presets (which updates the symlink) doesn't cause the sym link to be re-evaluated)

**Comment by decimad** [comment]
I would actually like to understand the synthesized path build system use-case that requires symlinks to be resolved better, because it just sounds so broken by design.

Also, if someone symlinks the same file twice into one workspace, and both files are also part of the same project, so that clangd struggles with ambiguity, then maybe he should think of a better strategy.

**Comment by JVApen** [comment]
Agreed. Let me already explain my use-case, as it sounds similar. (And doesn't require any changes to clangd or the extension)

```
root
|- source
|- build
   |- preset1
      |- codegen
   |- preset2 
      |- codegen
```

When creating an index file using preset1, all locations for the generated code in the index file are `build/preset1/codegen`.
For a user only/mostly uses preset2 and that index file, a 'find symbol' will give the path for preset1 and will as such find either no file or out-of-date file after selecting the symbol.

As such, we have a dedicated index-preset which puts codegen in `build/codegen`. Whenever you configure another preset, it instead creates a symlink `build/codegen` referring to the codegen dir of that preset.

As such, the find symbol simply works and resolves the path to actual location.

Note that this only holds for actions can find symbols for locations which are not yet locally indexed.

Our compile_commands.json and build system always use the actual directory and never go via the symlink. Given that the build directory is a subdir of `root`, which is the workspace, these files do exist twice in the workspace.

As indicated before, this works perfectly for us without any changes, on Windows and Linux.

**Comment by oold** [comment]
Our project structure looks like this:
```
D:
|- src (canonical source folder)
   |- a
   |- b
   |- proj/c (workspace root folder)
      |- build/compile_commands.json
      |- src
         |- a (symlink to D:/src/a)
         |- b (symlink to D:/src/b)
```
Navigating to header files inside the editor jumps outside the workspace and breaks clangd for us. Sure, I can hard-code absolute (relative doesn't work) Windows paths in the workspace `settings.json`, but that's not a good solution at all. I think I could accept that if I could use relative paths or paths with relative components at the very least. Using `${workspaceFolder}` also breaks this solution since it returns the path with lowercase drive letter (why does it even do that?) and, of course, can't use it for the target path because that would require relative path components. A fix for this soon-ish would be much appreciated.

**Comment by decimad** [comment]
> Agreed. Let me already explain my use-case, as it sounds similar. (And doesn't require any changes to clangd or the extension)
> 
> ```
> root
> |- source
> |- build
>    |- preset1
>       |- codegen
>    |- preset2 
>       |- codegen
> ```
> 

Okay, so what would break is that clangd would not notice the content/timestamps change when you swap the symlink, am I understanding the concerns correctly? Would maybe a "smart" cache that tracks symlinks and invalidates children when they change also meet the requirements? Without resolving the paths?

**Comment by decimad** [comment]
I find it intolerable that a tool forces symlink crap for the lack of a functionality to sort this out differently, and thereby preventing the user's use case of symlinks.
